src/shared/navgen/nav.cpp

```diff
@@ -40,13 +40,35 @@
 #include "shared/bg_gameplay.h"
 #include "navgen.h"
 
-static Log::Logger LOG( VM_STRING_PREFIX "navgen", "", Log::Level::NOTICE );
+// disable suppression in case multiple threads finishing together lead to a burst of output
+static auto LOG = Log::Logger( VM_STRING_PREFIX "navgen", "", Log::Level::NOTICE ).WithoutSuppression();
 
-void UnvContext::doLog(const rcLogCategory /*category*/, const char* msg, const int /*len*/)
+void UnvContext::doLog(rcLogCategory category, const char* msg, int len)
 {
-	if ( m_logEnabled )
+	std::string line(msg, len);
+	switch ( category )
 	{
-		LOG.Notice(msg);
+		case RC_LOG_WARNING:
+		case RC_LOG_ERROR:
+			RunOnMainThread( [line] { LOG.Warn( line ); } );
+			break;
+		case RC_LOG_PROGRESS:
+		default:
+			RunOnMainThread( [line] { LOG.Notice( line ); } );
+			break;
+	}
+}
+
+void UnvContext::RunOnMainThread( std::function<void()> f )
+{
+	mainThreadTasks_.push_back( std::move( f ) );
+}
+
+void UnvContext::DoMainThreadTasks()
+{
+	for ( auto &f : mainThreadTasks_ )
```

Comment:
Does this need to be cleared at the end of the function? If not, please comment.

Reply:
The idea is to only call it once. I think there is a comment elsewhere about that but I went ahead and added a `clear()` just to avoid suspicious appearances.